enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,381 @@
+| Only persistent volumes are necessary             | `microshift`, `microshift-storage-topolvm` |                                     |
+
+
+The packages for networking and storage would be different in upstrean, containing
+references to unrestricted (OKD) container images, removing the need for a pull secret.
+
+Reuse the kustomize loader so system and user components use the mechanism,
+extending its behaviour so binary plugins can be used in this content.
+
+
+`/usr/lib/microshift/components.d/004-multus/plugin`
+`/usr/lib/microshift/components.d/005-networking-ovnk/plugin`
+`/usr/lib/microshift/components.d/010-storage-topolvm/plugin`
+
+  MicroShift component manager would scan the `/usr/lib/microshift/components.d/`
+  directory executing the component plugins in alphabetical order.
+
+  `plugin` is an executable (in most cases a bash script will be the simplest
+  solution), using go binaries would be an option when complex structures
+  or connectivity back to the MicroShift API is required.
+
+  Today for the CNI/CSI components there is no wait or readiness conditions,
+  if this needed to be implemented in the future it could be implemented in
+  the `plugin`.
+
+The following changes would be introduced to packaging:
+
+* Convert `microshift` into a `microshift-core`, which does not depend
+  on `microshift-networking-ovnk`or `microshift-storage-topolvm`.
+
+* The `microshift-networking-ovnk` package would include the necessary
+  `/usr/lib/microshift/components.d/05-networking-ovnk/plugin` and collateral
+  yaml files, including CRDs.
+
+* The `microshift-storage-topolvm` package would include the necessary
+  `/usr/lib/microshift/components.d/10-storage-topolvm/plugin` and any
+   collateral yamls, including crds.
+
+* The rebase script should be updated to extract those core components
+  as and update the plugin or kustomize yamls. At some point each component
+  could implement its own rebase script.
+
+* Incompatible component rpms should be impossible to install together.
+  i.e. `microshift-networking-bridge` plus `microshift-networking-ovnk`,
+  this can be handled with Conflict spec clauses. (This is just an example,
+  since microshift-networking-bridge would not be necessary as it's the
+  default for container-networking)
+
+* Systemd services of the then optional networking components (like `microshift-ovs-init`)
+  should still start before the `microshift` systemd service, but `microshift` cannot
+  depend on `microshift-ovs-init` anymore. This can be acomplished by using the
+  Before=microshift clause in `microshift-ovs-init.service` and enabling this service
+  during the `microshift-networking-ovnk` sub-package install.
+  An alternative to this can be making `microshift-ovs-init` part of the plugin.
+
+#### Plugin calling convention
+
+The `plugin` binaries receive configuration details via environment variables, using the same variable format
+MicroShift already receives (see MicroShift's howto_config.md):
```

Comment:
IIRC, there's the idea to potentially remove configuration via these env vars from MicroShift (not proposing this, just noting). Also, the naming may need reviewing anyway, as we changed to a different user-facing config schema.

Reply:
The env var is being removed in the PR: https://github.com/openshift/microshift/pull/1384